plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/CabalAdd.hs

```diff
@@ -0,0 +1,272 @@
+      LogCalledCabalAddCodeAction -> "The CabalAdd CodeAction is called"
+      LogCalledCabalAddCommand params -> "Called CabalAdd command with:\n" Logger.<+> Logger.pretty params
+      LogCreatedEdit edit -> "Created inplace edit:\n" Logger.<+> Logger.pretty edit
+      LogExecutedCommand -> "Executed CabalAdd command"
+
+cabalAddCommand :: IsString p => p
+cabalAddCommand = "cabalAdd"
+
+data CabalAddCommandParams =
+     CabalAddCommandParams { cabalPath   :: FilePath
+                           , verTxtDocId :: VersionedTextDocumentIdentifier
+                           , buildTarget :: Maybe String
+                           , dependency  :: T.Text
+                           , version     :: Maybe T.Text
+                           }
+  deriving (Generic, Show)
+  deriving anyclass (FromJSON, ToJSON)
+
+instance Logger.Pretty CabalAddCommandParams where
+  pretty CabalAddCommandParams{..} =
+    "CabalAdd parameters:\n" Logger.<+>
+    "| cabal path: " Logger.<+> Logger.pretty cabalPath Logger.<+> "\n" Logger.<+>
+    "| target: " Logger.<+> Logger.pretty buildTarget Logger.<+> "\n" Logger.<+>
+    "| dependendency: " Logger.<+> Logger.pretty dependency Logger.<+> "\n" Logger.<+>
+    "| version: " Logger.<+> Logger.pretty version Logger.<+> "\n"
+
+-- | Gives a code action that calls the command,
+--   if a suggestion for a missing dependency is found.
+--   Disabled action if no cabal files given.
+--   Conducts IO action on a cabal file to find build targets.
+addDependencySuggestCodeAction
+  :: Logger.Recorder (Logger.WithPriority Log)
+  -> PluginId
+  -> VersionedTextDocumentIdentifier -- ^ Cabal's versioned text identifier
+  -> Int -- ^ Maximum number of suggestions to return
+  -> Diagnostic -- ^ Diagnostic from a code action
+  -> FilePath -- ^ Path to the haskell file
+  -> FilePath -- ^ Path to the cabal file
+  -> GenericPackageDescription
+  -> IO [CodeAction]
+addDependencySuggestCodeAction recorder plId verTxtDocId maxCompletions diag haskellFilePath cabalFilePath gpd = do
+    buildTargets <- liftIO $ getBuildTargets gpd cabalFilePath haskellFilePath
+    Logger.logWith recorder Logger.Info LogCalledCabalAddCodeAction
+    case buildTargets of
+      [] -> pure $ mkCodeAction cabalFilePath Nothing <$> hiddenPackageSuggestion maxCompletions (_message diag)
+      targets -> pure $ concat [mkCodeAction cabalFilePath (Just $ buildTargetToStringRepr target) <$>
+                          hiddenPackageSuggestion maxCompletions (_message diag) | target <- targets]
+  where
+    buildTargetToStringRepr target = render $ pretty $ buildTargetComponentName target
+
+    getBuildTargets :: GenericPackageDescription -> FilePath -> FilePath -> IO [BuildTarget]
+    getBuildTargets gpd cabalFilePath haskellFilePath = do
+      let haskellFileRelativePath = makeRelative (dropFileName cabalFilePath) haskellFilePath
+      readBuildTargets (verboseNoStderr silent) (flattenPackageDescription gpd) [haskellFileRelativePath]
```

Comment:
Comment please why you call `makeRelative` here! I know you know, but the next developer won't know ;)

Reply:
Forgot about this hurdle. I will make notes about it.